group/assignment6/move_cars.py

```diff
@@ -0,0 +1,59 @@
+def generate_moves(start, final):
+    """ ith element of start start is the car currently in the ith space
+    ith element of final start is the car that should be in ith space
+    empty space is 0
+    yields optimal sequence of moves in format (car number, old space, new space) """
+
+    # validation of the input
```

Comment:
This should probably be a separate function.
I think I remember something about each part of the code needing to have a separate responsibility?

It also seems logical to have validation be optional. And the validation for opt3 is going to be, mmm, less nontrivial, so makes sense to keep it separate from the beginning and have the opt3 validation reuse this piece

Reply:
I merged "car_positions" with "generate_moves" just because "car_positions" is called only once from "generate_moves".  So this merging makes the code shorter. I made validation the part of "generate_moves" for the same reason.
But I do not insist, because, yes, more functions can be more readable and testable, from the other side.
Let's see what others will say about separating code into logical blocks. I think that different approaches can exist there.

Reply:
having a different validation function could be a good idea. Easy to read and test 👍